packages/api/src/store/queue.ts

```diff
@@ -0,0 +1,155 @@
+  async handleMsg(msg: Notification) {
+    // emit it here
+    console.log("got NOTIFICATION: ", msg.channel, msg.payload);
+  }
+  // name: string;
+  // constructor({ db, schema }) {
+  //   super({ db, schema });
+  //   this.db = db;
+  //   this.schema = schema;
+  //   this.name = schema.queue;
+  // }
+
+  // get next event in queue
+  async pop(processFunc: Function): Promise<Queue> {
+    let res: QueryResult<DBLegacyObject>;
+    await this.client.query("BEGIN");
+    res = await this.client.query(
+      sql`SELECT data FROM `
+        .append(this.name)
+        .append(
+          sql` WHERE data->>'isConsumed' = 'false' AND data->>'status' = 'pending' LIMIT 1 FOR UPDATE SKIP LOCKED`.setName(
+            `${this.name}_next_event`
+          )
+        )
+    );
+
+    if (res.rowCount < 1) {
+      return null;
+    }
+
+    let originalData = res.rows[0].data;
+    // this allows us to pass a function that will process the event within the lock
+    if (processFunc) {
+      try {
+        await processFunc(originalData as Queue);
+      } catch (error) {
+        console.log("Msg queue process function error: ", error);
+        await this.client.query("ROLLBACK;");
+        throw error; // throwing the error till we can handle it better
+      }
+      res.rows[0].data.isConsumed = true;
+      res.rows[0].data.modifiedAt = Date.now();
+      await this.client.query(
+        sql`UPDATE queue SET data = data || ${res.rows[0].data} `.append(
+          ` WHERE id = '${res.rows[0].data.id}'`
+        )
+      );
+    } else {
+      res.rows[0].data.status = "processing";
+      res.rows[0].data.modifiedAt = Date.now();
+      console.log("id: ", res.rows);
+      await this.client.query(
+        sql`UPDATE queue SET data = data || ${res.rows[0].data} `.append(
+          ` WHERE id = '${res.rows[0].data.id}'`
+        )
+      );
+    }
+
+    await this.client.query("COMMIT;");
+    logger.debug(`MsgQueue consuming ${res.rows[0].id}`);
+    return originalData as Queue;
+  }
+
+  async emit(doc: Queue): Promise<Queue> {
+    if (!doc.status) {
+      doc.status = "pending";
+    }
+
+    try {
+      await this.client.query("BEGIN;");
+      await this.client.query(
+        `INSERT INTO ${this.name} VALUES ($1, $2)`, //p
+        [doc.id, JSON.stringify(doc)] //p
+      );
+
+      if (this.listener) {
+        // emit it via notifications
+        console.log("emitting NOTIFY");
+        await this.client.query(`NOTIFY ${this.channel}, '${doc.id}';`);
+      }
+      this.client.query("COMMIT;");
+    } catch (e) {
+      if (e.message.includes("duplicate key value")) {
+        throw new BadRequestError(e.detail);
+      }
+      throw e;
+    }
+    logger.debug(`MsgQueue emitting ${doc.id}`);
+    return doc;
+  }
+
+  async updateMsg(doc: Queue): Promise<Queue> {
+    doc.modifiedAt = Date.now();
+    await this.client.query(
+      sql`UPDATE queue SET data = data || ${doc} `.append(
+        ` WHERE id = '${doc.id}'`
```

Comment:
I think we'd want this to avoid injection? Want to make sure the id is parameterized
```suggestion
        sql` WHERE id = '${doc.id}'`
```